src/inbox/mod_inbox_rdbms_async.erl

```diff
@@ -0,0 +1,276 @@
+
+-spec reset_unread(HosType :: mongooseim:host_type(),
+                   InboxEntryKey :: mod_inbox:entry_key(),
+                   MsgId :: binary() | undefined) -> mod_inbox:write_res().
+reset_unread(HostType, Entry, MsgId) ->
+    PoolName = mongoose_async_pools:pool_name(HostType, inbox),
+    Params = {reset_unread, Entry, MsgId},
+    wpool:cast(PoolName, {task, Params}, {hash_worker, Entry}).
+
+-spec remove_inbox_row(HostType :: mongooseim:host_type(),
+                       InboxEntryKey :: mod_inbox:entry_key()) -> mod_inbox:write_res().
+remove_inbox_row(HostType, Entry) ->
+    PoolName = mongoose_async_pools:pool_name(HostType, inbox),
+    Params = {remove_inbox_row, Entry},
+    wpool:cast(PoolName, {task, Params}, {hash_worker, Entry}).
+
+%% synchronous callbacks
+-spec get_inbox(mongooseim:host_type(), jid:luser(), jid:lserver(), mod_inbox:get_inbox_params()) ->
+    get_inbox_res().
+get_inbox(HostType, LUser, LServer, Params) ->
+    mod_inbox_rdbms:get_inbox(HostType, LUser, LServer, Params).
+
+-spec get_inbox_unread(mongooseim:host_type(), mod_inbox:entry_key()) ->
+    {ok, integer()}.
+get_inbox_unread(HostType, Entry) ->
+    mod_inbox_rdbms:get_inbox_unread(HostType, Entry).
+
+-spec remove_domain(mongooseim:host_type(), jid:lserver()) -> ok.
+remove_domain(HostType, LServer) ->
+    mod_inbox_rdbms:remove_domain(HostType, LServer).
+
+-spec clear_inbox(mongooseim:host_type(), jid:luser(), jid:lserver()) ->
+    mod_inbox:write_res().
+clear_inbox(HostType, LUser, LServer) ->
+    mod_inbox_rdbms:clear_inbox(HostType, LUser, LServer).
+
+-spec get_entry_properties(mongooseim:host_type(), mod_inbox:entry_key()) ->
+    entry_properties() | nil().
+get_entry_properties(HostType, Entry) ->
+    mod_inbox_rdbms:get_entry_properties(HostType, Entry).
+
+-spec set_entry_properties(mongooseim:host_type(),
+                           mod_inbox:entry_key(),
+                           entry_properties()) ->
+    entry_properties() | {error, binary()}.
+set_entry_properties(HostType, Entry, Properties) ->
+    mod_inbox_rdbms:set_entry_properties(HostType, Entry, Properties).
+
+%% Internal functions
+
+flush_one(HostType, {set_inbox, Entry, Content, Count, MsgId, Timestamp}) ->
+    mod_inbox_rdbms:set_inbox(HostType, Entry, Content, Count, MsgId, Timestamp);
+flush_one(HostType, {set_inbox_incr_unread, Entry, Content, MsgId, Timestamp, Incrs}) ->
+    mod_inbox_rdbms:set_inbox_incr_unread(HostType, Entry, Content, MsgId, Timestamp, Incrs);
+flush_one(HostType, {remove_inbox_row, Entry}) ->
+    mod_inbox_rdbms:remove_inbox_row(HostType, Entry);
+flush_one(HostType, {reset_unread, Entry, MsgId}) ->
+    mod_inbox_rdbms:reset_unread(HostType, Entry, MsgId).
+
+-spec aggregate(NewTask :: async_task(), CurrentlyAccumulatedTask :: async_task()) -> FinalTask :: term().
+
+%%% First task being processed, just take that one
+aggregate(
+  Task,
+  undefined) ->
+    Task;
```

Comment:
Minor, this style looks a bit awkward to me, I know it is supposed to be more readable, but for me it is harder to see what is an argument and what is function body.

```suggestion
aggregate(Task, undefined) ->
    Task;
```

Reply:
Ok, I'll keep both variables into the same line, unless the line turns out to be too long (will keep it =< 100 columns strictly), then I'll have to partition it somehow.